### PR TITLE
feat(connection): support connection collection delete & rename & add(drop)

### PR DESCRIPTION
**Pull request by oceanlvr**
### PR Checklist

If you have any questions, you can refer to the [Contributing Guide](https://github.com/emqx/MQTTX/blob/master/.github/CONTRIBUTING.md)

#### What is the current behavior?

Please describe the current behavior and link to a relevant issue.

#### Issue Number

#375 

#### What is the new behavior?

add rename and new collection menu.

https://user-images.githubusercontent.com/36698124/112271363-cd85cf80-8cb5-11eb-8d0f-46c5d7086a4e.mp4

#### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### Specific Instructions

Are there any specific instructions or things that should be known prior to reviewing?

### CODE REVIEW

**Comment by oceanlvr** [comment]
@ysfscream please hold this PR, I'm working in the process for the collections persistence.

**Comment by ysfscream** [comment]
> please hold this PR, I'm working in the process for the collections persistence. :)

OK, If this PR is not ready, you can convert it to a draft until you think it is ready to review.

**Comment by oceanlvr** [comment]
Support edit/new collection now.

TODO:

- [ ] support connection drag to collection persistently
- [ ] remove collection

**Comment by oceanlvr** [comment]
@ysfscream Hi, sorry for my long long delay, Connection collection is available now, I need to more test to do, PTAL.

currently support:

- [x] persistence collection to `lowdb`
- [x] persistence connection's collection information
- [x] support drag and drop collection/connection node
- [x] collection support multi children and show as a tree view 
- [x] support collapse collection 
- [x] added context menu which support add/delete/rename collection

**Comment by oceanlvr** [comment]
support origin connection model:
image

add `add-collection` icon

image

type `enter` or blur mouse:

image

support drag and drop:

image

result:

image

File: src/utils/getCollectionId.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default (): string => {
+  return `collection_${Math.random().toString(16).substr(1, 7)}` as string
```
Comment (ysfscream) [inline src/utils/getCollectionId.ts:2]: I think it's better to use `lodash-id` to generate.

https://github.com/typicode/lodash-id

File: src/views/connections/ConnectionsList.vue [hunk 2]
```diff
@@ -3,49 +3,59 @@
...
-          draggable
           highlight-current
+          default-expand-all
           :current-node-key="connectionId"
+          @node-drop="handleDrop"
           :allow-drop="allowDrop"
         >
           <span class="custom-tree-node" slot-scope="{ node, data }">
+            <div v-if="data.isEdit">
+              <!-- <el-input></el-input> -->
```
Comment (ysfscream) [inline src/views/connections/ConnectionsList.vue:46]: Please remove this useless comment.

File: src/views/connections/ConnectionsList.vue [hunk 3]
```diff
@@ -256,6 +377,114 @@ export default class ConnectionsList extends Vue {
...
+
+  private handleRenameCollection() {
+    if (this.selectedCollection) {
+      this.selectedCollection.isEdit = true
+    }
+    this.showCollectionsContextmenu = false
+  }
+
+  private mounted() {
+    this.$nextTick(() => {
```
Comment (ysfscream) [inline src/views/connections/ConnectionsList.vue:484]: Why use $nextTick here?

File: src/views/connections/ConnectionsList.vue [hunk 4]
```diff
@@ -256,6 +377,114 @@ export default class ConnectionsList extends Vue {
...
+    if (selectedCollection && selectedCollection.isCollection) {
+      const nodeParent: ConnectionModelTree[] | ConnectionModelTree = this.getTreeNodeParent(selectedCollection) as
+        | ConnectionModelTree[]
+        | ConnectionModelTree
+      let childrenNode: ConnectionModelTree[]
+      if (Array.isArray(nodeParent)) {
+        childrenNode = nodeParent
+      } else {
+        childrenNode = (nodeParent as ConnectionModelCollection).children
+      }
```
Comment (ysfscream) [inline src/views/connections/ConnectionsList.vue:454]: Is this better?
```js
const childrenNode: ConnectionModelTree[] = Array.isArray(nodeParent) ? nodeParent : (nodeParent as ConnectionModelCollection).children
```

File: src/views/connections/ConnectionsList.vue [hunk 5]
```diff
@@ -256,6 +377,114 @@ export default class ConnectionsList extends Vue {
...
+      isEdit: true,
+    })
+    this.flushCollectionChange()
+  }
+
+  private getTreeNodeParent(node: ConnectionModelTree): ConnectionModelTree | ConnectionModelTree[] | null {
+    let findRes = false
+    let res: ConnectionModelTree | ConnectionModelTree[] | null = null
+    const treeData = this.treeData
+    treeData.forEach((e: ConnectionModelTree) => {
```
Comment (ysfscream) [inline src/views/connections/ConnectionsList.vue:413]: Please rename `e` to `tree` or another.

File: src/views/connections/ConnectionsList.vue [hunk 6]
```diff
@@ -136,81 +169,142 @@ export default class ConnectionsList extends Vue {
...
-    //TODO: sort
-    return []
+  // flush the Collection change to local storage
+  private async flushCollectionChange() {
+    this.treeData = [...this.treeData]
+    const tree = _.cloneDeep(this.treeData)
+    // get Collection without connection children
+    const travelCollection = (treeElements: ConnectionModelTree[]) => {
+      const res = [] as ConnectionModelCollection[]
+      if (!treeElements || (treeElements && !treeElements.length)) {
```
Comment (0721Betty) [inline src/views/connections/ConnectionsList.vue:202]: treeElements && !treeElements.length ---> treeElements.length < 1

File: src/views/connections/ConnectionsList.vue [hunk 7]
```diff
@@ -136,81 +169,142 @@ export default class ConnectionsList extends Vue {
...
     }
     return true
   }
 
-  get listdata(): ConnectionModelTree[] | [] {
-    //TODO: sort
-    return []
+  // flush the Collection change to local storage
+  private async flushCollectionChange() {
+    this.treeData = [...this.treeData]
```
Comment (0721Betty) [inline src/views/connections/ConnectionsList.vue:197]: What is the purpose of this statement？

File: src/views/connections/ConnectionsList.vue [hunk 8]
```diff
@@ -136,81 +169,142 @@ export default class ConnectionsList extends Vue {
...
+    }
+    if (!draggingNode.data.isCollection) {
+      switch (position) {
+        case 'inner':
+          draggingNode.data.collectionId = dropNode.data.id
+          updateConnectionCollectionId(draggingNode.data.id, dropNode.data.id)
+          break
+        default:
+          if (!dropNode.parent) return
+          if (Array.isArray(dropNode.parent.data)) {
```
Comment (0721Betty) [inline src/views/connections/ConnectionsList.vue:241]: use condition (ternary) operator?

**Review by 0721Betty — APPROVED** [review]

**Review by ysfscream — APPROVED** [review]

**Pull request merged**